### PR TITLE
RigidMotion and the interface DynamicFrame

### PR DESCRIPTION
**Pull request by eggrobin**
`DynamicFrame` will supersede or underlie the `first` part of `Transforms`

### CODE REVIEW

File: physics/dynamic_frame.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+using geometry::Instant;
+using geometry::OrthogonalMap;
+using geometry::Position;
+using geometry::Vector;
+using quantities::Acceleration;
+using quantities::Length;
+
+namespace physics {
+
+// TODO(egg): AffineMap needs an accessor for its linear map.
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: Remove.

File: physics/dynamic_frame.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  friend RigidMotion<From, To> operator*(
+      RigidMotion<Through, To> const& left,
+      RigidMotion<From, Through> const& right);
+
+ private:
+  RigidTransformation<FromFrame, ToFrame> rigid_transformation_;
+  // d/dt rigid_transformation(basis of FromFrame). The positively oriented
+  // orthogonal bases of |ToFrame| are acted upon faithfully and transitively by
+  // SO(ToFrame), so this lies in the tangent space, i.e., the Lie algebra
+  // ????(ToFrame) = ToFrame ? ToFrame.
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: Why the ????

File: physics/dynamic_frame.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  Velocity<ToFrame> translation_;
+};
+
+template<typename FromFrame, typename ThroughFrame, typename ToFrame>
+RigidMotion<FromFrame, ToFrame> operator*(
+    RigidMotion<ThroughFrame, ToFrame> const& left,
+    RigidMotion<FromFrame, ThroughFrame> const& right);
+
+// The definition of a reference frame |ThisFrame| in arbitrary motion with
+// respect to the inertial reference frame |BaseFrame|.
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: If `BaseFrame` has to be inertial maybe we could call it `InertialFrame`?  The name would be a bit more informative, I think.

File: physics/dynamic_frame.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+template<typename BaseFrame, typename ThisFrame>
+class DynamicFrame {
+ public:
+  static_assert(BaseFrame::is_inertial, "BaseFrame must be inertial");
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: Move before `public:`.

File: physics/dynamic_frame.hpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  virtual ~DynamicFrame() = 0;
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: This is odd.  Will the subclasses be able to destroy?

File: physics/dynamic_frame_body.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "physics/dynamic_frame.hpp"
+
+namespace principia {
+namespace physics {
```
Comment (pleroy) [inline physics/dynamic_frame_body.hpp]: Blank line after.

File: physics/dynamic_frame_body.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+template <typename FromFrame, typename ToFrame>
```
Comment (pleroy) [inline physics/dynamic_frame_body.hpp]: No space before `<`.  I think that's what we generally do, and that's what you did in the declaration.

File: physics/dynamic_frame_body.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+OrthogonalMap<FromFrame, ToFrame> const&
+RigidMotion<FromFrame, ToFrame>::orthogonal_map() const {
+  return rigid_transformation_.linear_map();
+}
+
+template <typename FromFrame, typename ToFrame>
+DegreesOfFreedom<ToFrame> RigidMotion<FromFrame, ToFrame>::operator()(
+    DegreesOfFreedom<FromFrame> const& degrees_of_freedom) const {
+  Position<ToFrame> const result_position =
```
Comment (pleroy) [inline physics/dynamic_frame_body.hpp]: Is this intermediate variable needed?  I would just put the evaluation at line 33.

File: physics/dynamic_frame_body.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      rigid_transformation_(degrees_of_freedom.position());
+  return {result_position,
+          orthogonal_map()(degrees_of_freedom.velocity()) + translation_ +
+              rotation_ * orthogonal_map()(degrees_of_freedom.position() -
+                                           FromFrame::origin) / Radian};
+}
+
+template <typename FromFrame, typename ToFrame>
+RigidMotion<ToFrame, FromFrame> RigidMotion<FromFrame, ToFrame>::Inverse()
+    const {
```
Comment (pleroy) [inline physics/dynamic_frame_body.hpp]: Preference for breaking before the method name.

File: physics/dynamic_frame_test.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
+﻿#include "physics/dynamic_frame.hpp"
+
+#include "geometry/frame.hpp"
+#include "geometry/permutation.hpp"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "quantities/numbers.hpp"
+#include "quantities/named_quantities.hpp"
```
Comment (pleroy) [inline physics/dynamic_frame_test.cpp]: Alphabetize.

File: physics/dynamic_frame.hpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+template <typename FromFrame, typename ToFrame>
+using RigidTransformation =
+    AffineMap<FromFrame, ToFrame, Length, OrthogonalMap>;
+
+// The instantaneous motion of |ToFrame| with respect to |FromFrame|.
+// This is the derivative of a |RigidTransformation<FromFrame, ToFrame>|.
+// In order to invert, the |RigidTransformation| is needed, and we need its
+// linear part anyway, so we store it (and we forward its action on positions).
+template<typename FromFrame, typename ToFrame>
+class RigidMotion {
```
Comment (pleroy) [inline physics/dynamic_frame.hpp]: I would prefer to split this file in two parts: `RigidTransformation` + `RigidMotion` in one file, `rigid_motion.hpp`.  And the interface `DynamicFrame` in `dynamic_frame.hpp`.  1 file == 1 class == good.  As written, `dynamic_frame_test.cpp` doesn't test `DynamicFrame` at all (which is fine, an interface doesn't need tests).

File: physics/dynamic_frame.hpp (hunk 1 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Done.

File: physics/dynamic_frame.hpp (hunk 2 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Bad encoding, fixed.

File: physics/dynamic_frame.hpp (hunk 3 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Done.

File: physics/dynamic_frame.hpp (hunk 5 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Maybe, but let's just default it.

File: physics/dynamic_frame.hpp (hunk 4 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Done.

File: physics/dynamic_frame_body.hpp (hunk 6 above)
Comment (eggrobin) [inline physics/dynamic_frame_body.hpp]: Done.

File: physics/dynamic_frame_body.hpp (hunk 7 above)
Comment (eggrobin) [inline physics/dynamic_frame_body.hpp]: Done.

File: physics/dynamic_frame_body.hpp (hunk 8 above)
Comment (eggrobin) [inline physics/dynamic_frame_body.hpp]: Done.

File: physics/dynamic_frame_body.hpp (hunk 9 above)
Comment (eggrobin) [inline physics/dynamic_frame_body.hpp]: Done.

File: physics/dynamic_frame_test.cpp (hunk 10 above)
Comment (eggrobin) [inline physics/dynamic_frame_test.cpp]: Done.

File: physics/dynamic_frame.hpp (hunk 11 above)
Comment (eggrobin) [inline physics/dynamic_frame.hpp]: Done.

**Pull request merged**